crates/ide-assists/src/handlers/generate_mut_trait_impl.rs

```diff
@@ -0,0 +1,195 @@
+use syntax::{
+    ast::{self, make},
+    ted, AstNode,
+};
+
+use crate::{AssistContext, AssistId, AssistKind, Assists};
+
+// Generate proper `index_mut` method body refer to `index` method body may impossible due to the unpredicable case [#15581].
+// Here just leave the `index_mut` method body be same as `index` method body, user can modify it manually to meet their need.
+
+// Assist: generate_mut_trait_impl
+//
+// Adds a IndexMut impl from the `Index` trait.
+//
+// ```
+// pub enum Axis { X = 0, Y = 1, Z = 2 }
+//
+// impl<T> Index$0<Axis> for [T; 3] {
+//     type Output = T;
+//
+//     fn index(&self, index: Axis) -> &Self::Output {
+//         &self[index as usize]
+//     }
+// }
+// ```
+// ->
+// ```
+// pub enum Axis { X = 0, Y = 1, Z = 2 }
+//
+// $0impl<T> IndexMut<Axis> for [T; 3] {
+//     fn index_mut(&mut self, index: Axis) -> &mut Self::Output {
+//         &self[index as usize]
+//     }
+// }
+//
+// impl<T> Index<Axis> for [T; 3] {
+//     type Output = T;
+//
+//     fn index(&self, index: Axis) -> &Self::Output {
+//         &self[index as usize]
+//     }
+// }
+// ```
+pub(crate) fn generate_mut_trait_impl(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let impl_def = ctx.find_node_at_offset::<ast::Impl>()?.clone_for_update();
+
+    let trait_ = impl_def.trait_()?;
+    if let ast::Type::PathType(trait_type) = trait_.clone() {
+        let trait_name = trait_type.path()?.segment()?.name_ref()?.to_string();
+        if trait_name != "Index" {
+            return None;
+        }
+    }
+
+    // Index -> IndexMut
+    let index_trait = impl_def
```

Comment:
We need to check right after we figured out what the `ast::Type` is, whether `Index` actually comes from `std::ops::Index`. See [this](https://github.com/rust-lang/rust-analyzer/blob/69432287cbaa0cd69dd7219aa0c6c74e44dbe7e7/crates/ide-assists/src/handlers/into_to_qualified_from.rs#L50) as a reference.